Objects/longobject.c

```diff
@@ -1270,8 +1276,9 @@ PyLong_FromSize_t(size_t ival)
     size_t t;
     int ndigits = 0;
 
-    if (ival < PyLong_BASE)
-        return PyLong_FromLong((long)ival);
```

Comment:
How about making this just call `PyLong_FromUnsignedLong` instead?

... Ah, it looks like `PyLong_FromUnsignedLong` does the same thing. Well: how about applying this optimization to `PyLong_FromUnsignedLong` instead of `PyLong_FromSize_t`, and making this just call `PyLong_FromUnsignedLong` instead? :wink: 

The usual practice across the codebase is to use either `PyLong_FromLong` or `PyLong_FromUnsignedLong` whenever they'll do, so there are a lot more calls to `PyLong_FromUnsignedLong` than to `PyLong_FromSize_t`. Which means the optimization will apply much more broadly if it's present there.

Reply:
> Which means the optimization will apply much more broadly if it's present there.

I'm curious as to how much of an optimization this would lead to for the performance of ``PyLong_FromUnsignedLong()``. There are approximately 81 calls to the function, (according to ``git grep -E "PyLong_FromUnsignedLong\(.+\)" -- '*.c' | wc -l``) so even a minor optimization could make a significant difference.

Reply:
Probably I'll refactor PyLong_From*() functions like in #15457 later, but I don't want to make any other changes in this PR because it blocks me from further work, specifically from replacing `PyLong_FromSsize_t()` with `PyLong_FromSsize_t()` in the places where it makes sense.

Reply:
Well, a version of this PR that applies to the ...UnsignedLong function is literally one line bigger than this version is. (Apply the same diff to that function instead of this one; and in this one, just s/Long/UnsignedLong/.) And it makes the change more logically coherent. So I'm not sure that making that revision will make it any slower to get the PR merged; it could make it faster.

That other PR #15457 looks much more complex.

> replacing PyLong_FromSsize_t() with PyLong_FromSsize_t()

I think you meant something else here. ;-) I'm curious exactly what.

Reply:
I meant `PyLong_FromSsize_t()` with `PyLong_FromSize_t()`.

Reply:
> Apply the same diff to that function instead of this one; and in this one, just s/Long/UnsignedLong/

`if (ival < PyLong_BASE)` makes `PyLong_FromSsize_t()` slower than it is in this PR.

Reply:
> `if (ival < PyLong_BASE)` makes `PyLong_FromSsize_t()` slower than it is in this PR.

Hmm, I didn't mean to suggest doing anything at all to `PyLong_FromSsize_t`; so I think there must have been a misunderstanding. I'll try again, with more explicitness. (I wrote my previous message from a phone, which made typing harder. Back at a nice keyboard now.)

The current version of the PR has this diff:
```
-    if (ival < PyLong_BASE)
-        return PyLong_FromLong((long)ival);
+    if (is_small_uint(ival)) {
+        return get_small_int(ival);
+    }
```
inside the implementation of `PyLong_FromSize_t`.

I think it would be better to (a) apply essentially the exact same diff:
```
-    if (ival < PyLong_BASE)
-        return PyLong_FromLong(ival);
+    if (is_small_uint(ival)) {
+        return get_small_int(ival);
+    }
```
but inside the implementation of `PyLong_UnsignedLong`. Then (b) inside the implementation of this function `PyLong_FromSize_t`, apply instead this diff:
```
     if (ival < PyLong_BASE)
-        return PyLong_FromLong((long)ival);
+        return PyLong_FromUnsignedLong((unsigned long)ival);
```

This will make the optimization apply more broadly. It will make the logic of the change better fit with how we generally use the `PyLong_From*` functions.

And the cost is one additional line changed in the diff, and zero additional complexity in the resulting code.

Reply:
> Hmm, I didn't mean to suggest doing anything at all to PyLong_FromSsize_t

I mixed up them again, actually I meant `PyLong_FromSize_t()`.

Reply:
> I mixed up them again, actually I meant `PyLong_FromSize_t()`.

Got it.

I suggested pointing at `PyLong_FromUnsignedLong` because your original explanation (in the issue thread) for why this was an optimization was that `PyLong_FromLong` needs to handle the sign, and `PyLong_FromUnsignedLong` would suffice to avoid that.

Here's a variation: take that same 2-line diff that replaces `if (ival < PyLong_BASE)` and `PyLong_FromLong` with `is_small_int` and `get_small_int`... and apply it to *both* of `PyLong_FromUnsignedLong` and `PyLong_FromSize_t`.

That will again make the optimization apply more broadly; and it'll keep these two functions closely parallel, as one would naturally expect them to be.

The cost this time is two additional lines changed in the diff (three, counting the close-brace). And it actually reduces complexity in the resulting code, because it doesn't leave readers having to wonder why the constructors taking `unsigned long` on the one hand and `size_t` on the other -- which are even the same type on common platforms -- have different logic.